### PR TITLE
Cube build stage3 - JP-2096

### PR DESCRIPTION
**Pull request by jemorrison**
**Description**

Speeding up cube_build using numba.  Numba was installed using 'pip'
The purpose of this PR is to speed up cube_build using numba. Some of the modules in ifu_cube.py were moved out
of the class and made independent routines. Some of these routines were broken up into simpler routines. In this process the  weighting by the Miri psi option was removed. It is not being used and it was getting cumbersome to it as an option. Removing Miri ps weighting as an option also means that the resolution reference files is not longer needed. 

Including this PR in the JW pipeline requires numba to be added. But so far there seems little down side to including numba. It is fast, stable and easy to use.

### CODE REVIEW

**Comment by jemorrison** [comment]
PR ready for review.

**Comment by jemorrison** [comment]
See information on JP-2096 on numba and speed tests D Law has run with the code.

**Comment by jemorrison** [comment]
Once it is approved that numba can be used in JWST Pipeline I can make further speed improvements in the blotting routines and have them use numba. I am going to hold off on that until it is confirmed numba is fine to use the jwst pipeline

**Comment by hbushouse** [comment]
@jdavies-st or @eslavich any ideas what's killing the CI test "CI/Installed package with --pyargs"?

**Comment by jdavies-st** [comment]
If you search the log for "FAILURES"  you can see it is having trouble `asdf.open()` on a downloaded CRDS NIRISS WFSS reference file.

**Comment by jdavies-st** [comment]
We should be opening those reffiles in a cleaner way, i.e. with a `with` context manager so that if they fail to load, we don't get a cascade of errors.

**Comment by jemorrison** [comment]
@jdavies-st @nden if you have time now could you look this over or suggest someone else to look at it.
It is using numba to speed up cube_build

**Comment by jemorrison** [comment]
@hbushouse @nden @mcara 
This PR is ready for review. This PR pulled out the computational  python modules and converted them to c routines. There are now two c extensions 1. creating IFU cubes on sky 2. creating IFU cubes in detector plane.
I made a common set of c functions that both c extensions call in  cube_utilts.c 
As noted above in the comments - an unused weighting function - miripsf - was removed as an option. 
The unit tests run and the regression tests also run (with a few differences due to improvements I made when creating c routines).

**Comment by drlaw1558** [comment]
Following the usual means of checking out a PR, I'm getting errors about being unable to find the C modules.  E.g.,
from .cube_match_sky import cube_wrapper
returns
No module named jwst.cube_build.cube_match_sky

I'm probably just doing something wrong- do I need to do anything special to compile the code inside the src/ directory first or should python do that automatically?

**Comment by jemorrison** [comment]
To compile the c code. In the top jwst directory - the directory with setup.py
Type pip install -e .
That compiles the c code that is defined in setup.py
Let me you if you have problems with that

File: jwst/cube_build/src/cube_match_internal.c
Comment (mcara) [inline jwst/cube_build/src/cube_match_internal.c:167]: I would suggest that 1) `mem_alloc` be renamed to something less generic and 2) use `calloc` in order to initialize arrays to 0. For example:
```suggestion
int alloc_flux_arrays(int nelem, double **fluxv, double **weightv, double **varv,  double **ifluxv) {
    const char *msg = "Couldn't allocate memory for output arrays.";
    
    // flux:
    if (!(*fluxv  = (double*)calloc(nelem, sizeof(double)))) {
        PyErr_SetString(PyExc_MemoryError, msg);
        goto failed_mem_alloc;
    }

    // weight:
    if (!(*weightv = (double*)calloc(nelem, sizeof(double)))) {
        PyErr_SetString(PyExc_MemoryError, msg);
        goto failed_mem_alloc;
    }

    // variance:
    if (!(*varv = (double*)calloc(nelem, sizeof(double)))) {
        PyErr_SetString(PyExc_MemoryError, msg);
        goto failed_mem_alloc;
    }
    // iflux
    if (!(*ifluxv = (double*)calloc(nelem, sizeof(double)))) {
        PyErr_SetString(PyExc_MemoryError, msg);
        goto failed_mem_alloc;
    }

    return 0;
    
failed_mem_alloc:
    free(*fluxv);
    free(*weightv);
    free(*varv);
    free(*ifluxv);
}

//_______________________________________________________________________
//  based on a 2-D drizzling type of algorithm find the contribution of
// each detector flux to the IFU cube
//_______________________________________________________________________

int match_detector_cube(int instrument, int naxis1, int naxis2, int nz, int npt, int ncube, int na,
			double crval_along, double cdelt_along, double crval3, double cdelt3,
			double *a1, double *a2,double *a3, double*a4,
			double *lam1, double *lam2, double*lam3, double *lam4,
			double *acoord, double *zcoord, int ss,
			double *pixel_flux, double *pixel_err,
			double **spaxel_flux, double **spaxel_weight, double **spaxel_var,double **spaxel_iflux){

  double *fluxv, *weightv, *varv;  // vectors for spaxel 
  double *ifluxv;  // vector for spaxel

  // allocate memory to hold output 
  if (alloc_flux_arrays(ncube, &fluxv, &weightv, &varv, &ifluxv)) return 1;
    
```

File: jwst/cube_build/src/cube_match_sky.c
Comment (mcara) [inline jwst/cube_build/src/cube_match_sky.c:108]: Isn't this the same function as in `cube_match_internal.c`? Maybe it should be placed in a separate file and then included in both modules?

File: jwst/cube_build/src/cube_match_sky.c
Comment (mcara) [inline jwst/cube_build/src/cube_match_sky.c:164]:
```suggestion
    const char *msg = "Couldn't allocate memory for output arrays.";

    if (!(*idqv = (int*)calloc(nelem, sizeof(int)))) {
        PyErr_SetString(PyExc_MemoryError, msg);
        return 1;
    }
```

File: jwst/cube_build/src/cube_match_sky.c
Comment (mcara) [inline jwst/cube_build/src/cube_match_sky.c:633]: Assuming you have switched to `calloc` as suggested above
```suggestion
    if (mem_alloc_dq(ncube, spaxel_dq)) return 1;
```

File: jwst/cube_build/src/cube_match_sky.c
Comment (mcara) [inline jwst/cube_build/src/cube_match_sky.c:624]: function name seems confusing/not informative.

File: jwst/cube_build/src/cube_match_sky.c
Comment (mcara) [inline jwst/cube_build/src/cube_match_sky.c:657]:
```suggestion
    int *idqv;  // int vector for spaxel
    if (mem_alloc_dq(ncube, &idqv)) return 1;
```

File: jwst/cube_build/src/cube_match_sky.c
Comment (mcara) [inline jwst/cube_build/src/cube_match_sky.c:755]:
```suggestion
  int *idqv;  // int vector for spaxel
  if (mem_alloc_dq(ncube, &idqv)) return 1;
```

File: jwst/cube_build/src/cube_match_sky.c
Comment (mcara) [inline jwst/cube_build/src/cube_match_sky.c:767]:
```suggestion
      int status; 
```

File: jwst/cube_build/src/cube_match_sky.c
Comment (mcara) [inline jwst/cube_build/src/cube_match_sky.c:833]: This can be simplified as in `cube_match_internal.c` if using `alloc_flux_arrays`.

File: jwst/cube_build/src/cube_match_sky.c
Comment (mcara) [inline jwst/cube_build/src/cube_match_sky.c:1000]: Again this can be simplified as in `cube_match_internal.c` if using `alloc_flux_arrays`.

File: jwst/cube_build/src/cube_match_internal.c
Comment (hbushouse) [inline jwst/cube_build/src/cube_match_internal.c:2]: typo: "could" -> "cloud"

File: jwst/cube_build/src/cube_match_internal.c
Comment (hbushouse) [inline jwst/cube_build/src/cube_match_internal.c:4]: typo? Do you mean "size of the spaxel *in* spatial coords?

File: jwst/cube_build/src/cube_match_internal.c
Comment (hbushouse) [inline jwst/cube_build/src/cube_match_internal.c:5]: typo? should "zcdelt3" be just "cdelt3"?

Comment (hbushouse) [reply jwst/cube_build/src/cube_match_internal.c:5]: Oh, and "cdetl1" should be "cdelt1"

File: jwst/cube_build/src/cube_match_internal.c
Comment (hbushouse) [inline jwst/cube_build/src/cube_match_internal.c:6]: do you really mean "the *e* modified shepard weighting" or is the "e" extraneous? And another instance of "clold" that should be "cloud".

File: jwst/cube_build/src/cube_match_sky.c
Comment (hbushouse) [inline jwst/cube_build/src/cube_match_sky.c:2]: "could" -> "cloud"

File: jwst/cube_build/src/cube_match_sky.c
Comment (hbushouse) [inline jwst/cube_build/src/cube_match_sky.c:5]: All the same typos as the same paragraph in the previous module (cdetl1, zcdelt3, ...)

File: setup.cfg
Comment (hbushouse) [inline setup.cfg:38]: Now that we're using C extensions instead of numba, can this be removed?

File: jwst/cube_build/blot_cube.py
Comment (jemorrison) [reply jwst/cube_build/blot_cube.py:33]: I am going to hold off making changes to blot_cube.py because I have another JP ticket to work just on blot cube after I get the c extensions in this PR committed. I will come back to these changes suggestions later this week.

File: jwst/cube_build/blot_cube_build.py
Comment (jemorrison) [reply jwst/cube_build/blot_cube_build.py:336]: it is for NIRSPEC. It can take several seconds per slice. Once we get it faster I will remove the debug timing

File: jwst/cube_build/cube_build_wcs_util.py
Comment (jemorrison) [reply jwst/cube_build/cube_build_wcs_util.py:129]: nice! I did not know that.

**Comment by jemorrison** [comment]
@jdavies-st @eslavich 
The PR failed on CI/Installed package with --pyargs 
What does the mean  ? I looked at the details but I am a bit lost how to fix it

File: setup.cfg
Comment (jemorrison) [reply setup.cfg]: yes I forgot I put that in

File: jwst/cube_build/src/cube_match_sky.c
Comment (jemorrison) [reply jwst/cube_build/src/cube_match_sky.c:5]: fixed typos

File: jwst/cube_build/src/cube_match_sky.c
Comment (jemorrison) [reply jwst/cube_build/src/cube_match_sky.c]: good suggestion. I moved alloc_flux_arrays to cube_utils.c

**Comment by jdavies-st** [comment]
It looks like it was getting a corrupted CRDS reference file.  I kicked off the CI again.  Hopefully it won't be corrupted this time.

File: jwst/cube_build/src/cube_match_internal.c
Comment (jemorrison) [reply jwst/cube_build/src/cube_match_internal.c]: fixed

File: jwst/cube_build/src/cube_match_internal.c
Comment (jemorrison) [reply jwst/cube_build/src/cube_match_internal.c]: fixed

File: jwst/cube_build/src/cube_match_internal.c
Comment (jemorrison) [reply jwst/cube_build/src/cube_match_internal.c]: fixed

File: jwst/cube_build/src/cube_match_sky.c
Comment (jemorrison) [reply jwst/cube_build/src/cube_match_sky.c]: done

File: jwst/cube_build/src/cube_match_sky.c
Comment (jemorrison) [reply jwst/cube_build/src/cube_match_sky.c]: done

File: jwst/cube_build/src/cube_match_sky.c
Comment (jemorrison) [reply jwst/cube_build/src/cube_match_sky.c]: done

File: jwst/cube_build/blot_cube_build.py
Comment (jemorrison) [reply jwst/cube_build/blot_cube_build.py:329]: holding off on blot changes - I have opened a separate ticket on blotting

**Comment by jemorrison** [comment]
I made the changes suggested in the review. I reran the regression tests and  they look good. There are some expected differences because I improved  the DQ flags for edge cases and I made a few other little improvements. 
I think this is ready to be merged

**Comment by jdavies-st** [comment]
FYI the doc build is failing because the new C module cannot be imported for the doc build.  @eslavich do you recall what the issue was for C extensions and doc builds?

**Comment by hbushouse** [comment]
We made a slight change in docs/conf.py or something like that to remove a line that was causing it to look for things in a parent directory. Look at https://github.com/spacetelescope/jwst/pull/6207/files

**Comment by drlaw1558** [comment]
I'm seeing some larger changes in the data cubes than I would have expected given this just changes the language not the algorithm; let me investigate further and get back to you.  Did anything change that should have affected the total cube FOV?

**Comment by jemorrison** [comment]
@drlaw1558  I tweaked the DQ flagging.  Hopefully improving it near boundaries. I also tweaked- for MIRI - how the wavelength range of valid data is determined. There was a small bug in the old code.

**Comment by eslavich** [comment]
We merged a fix for the doc build failure in https://github.com/spacetelescope/jwst/pull/6230, so a rebase ought to get that unstuck.

**Comment by drlaw1558** [comment]
Ok, disregard my last comment- the changes that I was seeing were due to something changing earlier in the pipeline (that I'll have to run down elsewhere), not cube build.  Performance looks good to me; when the test is set up properly SCI results from running spec3 in multiple different modes look identical before/after this change but with vastly improved runtimes.

**Review by drlaw1558 — APPROVED** [review]
Focusing entirely on results, the performance of this PR looks good to me.  Running some test cases with dithered exposures in multiple bands through a variety of different kinds of cube building I get identical SCI results before/after this change but with vastly improved runtimes.  DQ arrays look improved.

**Comment by jemorrison** [comment]
@hbushouse can we merge this PR ?

**Review by hbushouse — APPROVED** [review]

File: jwst/cube_build/src/cube_match_internal.c
Comment (nden) [inline jwst/cube_build/src/cube_match_internal.c:120]: @jemorrison I believe you need to declare the variables in the beginning of the file as certain compilers won't like the definition within the for loop.

File: jwst/cube_build/src/cube_match_internal.c
Comment (nden) [inline jwst/cube_build/src/cube_match_internal.c:138]: Same comment as above

File: jwst/cube_build/src/cube_match_internal.c
Comment (nden) [inline jwst/cube_build/src/cube_match_internal.c:165]: and here

File: jwst/cube_build/src/cube_match_internal.c
Comment (nden) [inline jwst/cube_build/src/cube_match_internal.c:168]: and here

**Comment by jemorrison** [comment]
@nden 
As I was moving all the definition to top of code. I see that I am I doing this:

nxy = nx * ny
int wave_slice_dq[nxy]

The code works fine doing this on my Mac. I am now thinking this may not be allowed - dynamically allocated the array. Should I change how I do this ?

**Pull request merged**
